core/system/src/export_table.c

```diff
@@ -83,16 +95,446 @@ static void thread_destroy(void * c)
         /* Release this slot. */
         context->allocator = NULL;
     } else {
-        HALT_ERROR(SANITY_CHECK_FAILED,
-            "thread context (%08x) is invalid!\n",
-            context);
+        /* This is a debug only assertion, not present in release builds, to
+         * prevent a malicious box from taking down the entire system by
+         * fiddling with one of its thread contexts or destroying another box's
+         * thread. */
+        DPRINTF("thread context (%08x) is invalid!\r\n", context);
+        assert(false);
+    }
+}
+
+/* Wake up all the potential handlers for this RPC target. Return number of
+ * handlers posted to. */
+static int wake_up_handlers_for_target(const TFN_Ptr function, int box_id)
+{
+    int num_posted = 0;
+
+    UvisorBoxIndex * index = (UvisorBoxIndex *) g_context_current_states[box_id].bss;
+    uvisor_pool_queue_t * fn_group_queue = &index->rpc_fn_group_queue->queue;
+    uvisor_rpc_fn_group_t * fn_group_array = index->rpc_fn_group_queue->fn_groups;
+
+    /* Wake up all known waiters for this function. Search for the function in
+     * all known function groups. We have to search through all function groups
+     * (not just those currently waiting for messages) because we want the RTOS
+     * to be able to pick the highest priority waiter to schedule to run. Some
+     * waiters will wake up and find they have nothing to do if a higher
+     * priority waiter already took care of handling the incoming RPC. */
+    uvisor_pool_slot_t slot;
+    slot = fn_group_queue->head;
+    while (slot < fn_group_queue->pool->num) {
+        /* Look for the function in this function group. */
+        uvisor_rpc_fn_group_t * fn_group = &fn_group_array[slot];
+
+        TFN_Ptr const * fn_ptr_array = fn_group->fn_ptr_array;
+        size_t i;
+
+        for (i = 0; i < fn_group->fn_count; i++) {
+            /* If function is found: */
+            if (fn_ptr_array[i] == function) {
+                /* Wake up the waiter. */
+                semaphore_post(&fn_group->semaphore);
+                ++num_posted;
+            }
+        }
+
+        slot = fn_group_queue->pool->management_array[slot].queued.next;
+    }
+
+    return num_posted;
+}
+
+static int callee_box_id(const TRPCGateway * gateway)
+{
+    return (uint32_t *)gateway->box_ptr - __uvisor_config.cfgtbl_ptr_start;
```

Comment:
@Patater Make sure that you only return valid box ID's within start/end of the box-id pointer list - also ensure alignment of 4. Make sure to return a invalid ID if the pointer is outside of the box list.